Mm input mask guidance (#1979)
uswds/uswds-site: _components/input-mask/guidance/implementation.md

{:.usa-content-list }

> The examples in the Preview section above are common input masks, but you can adapt the code to meet the needs of your project. In order to extend the functionality of the input mask, you must include a `placeholder` attribute as the visible input mask, using the underscore `_` to represent a character the user should enter. You may use punctuation literals like `[space]`, `-`, `/`, or `(` to help format your mask, but these literals will not be included in the final submitted value. _See a complete list of format characters below_.

Reviewer: Instead of giving directions for where to find the content, it would be better to provide a direct link to the section. To match format, it would also be good to have a content summary in bold. One suggestion:
```suggestion
**Customize the pattern.** The examples in the [component preview section]({{ site.baseurl }}/components/input-mask/#input-mask-preview-content) show common input mask patterns, but you can adapt the code to meet the needs of your project. In order to extend the functionality of the input mask, you must include a `placeholder` attribute as the visible input mask, using the underscore `_` to represent a character the user should enter. You may use punctuation literals like `[space]`, `-`, `/`, or `(` to help format your mask, but these literals will not be included in the final submitted value. A full list of formatting characters can be found in the [Available format characters table]({{ site.baseurl }}/components/input-mask/#available-format-characters).
```

- **Add a `mask` attribute to the input element.** The mask attribute is used to define the format of your mask. Use `9`s for numbers and `A`s for letters. For example, a typical SSN number format in the mask attribute should look like `999-99-9999`.
- **Add a `placeholder` attribute to the input element.** This indicates the mask format. Use underscores `_` for character input placeholders combined with any valid format characters (see below) to customize your input mask. e.g. A US telephone `placeholder` attribute may look like `(___) ___-____`
- **Add a `maxlength` attribute to the input element.** This will be used as the limit referenced in the message and for validation.
- **Set the attribute `type` to numeric if your mask is all numbers.** This will display the numeric keypad instead of the full keyboard.

{:.usa-content-list }
The following is an example of `input` element attributes for a Social Security Number mask:

{:.usa-content-list }
```
aria-describedby="ssn-hint ssn-hint-info"
class="usa-input usa-input-mask__field"
id="ssn"
inputmode="numeric"
mask="999-99-9999"
maxlength="11"
name="ssn"
placeholder="___-__-____"

> type="text"

Reviewer: This code sample does not match what is shown for the same pattern in the component code section. Can the two align? Alternatively, does the code need to be repeated?

```

> {:.usa-content-list }
> <h3>The input mask and alphanumeric characters</h3>

Reviewer: It should be safe to remove the `usa-content-list` class from the header elements. Additionally, can you replace the `<h3>` elements and replace them with `###`?
```suggestion
### The input mask and alphanumeric characters
```

{:.usa-content-list }

> Input mask examples include an alphanumeric version which can be customized to meet your needs. Include the number `9` for every number and a capitalized letter `A` for every required letter. We recommend changing inputmode to `text` so the full keyboard is available to your user.

Reviewer: ```suggestion
Input mask examples include an alphanumeric version which can be customized to meet your needs. Include the number `9` for every number and a capitalized letter `A` for every required letter. We recommend changing `inputmode` to `text` so the full keyboard is available to your user.
```

{:.usa-content-list }
If your input mask calls for a complicated pattern like _1EG4-TE5-MK73_, you could implement that by adding the following attributes and character formatting to your `input` element:

{:.usa-content-list }
```
aria-describedby="alphanumeric-hint alphanumeric-hint-info"
class="usa-input usa-input-mask__field"
forceupper="true"
id="alphanumeric"
mask="9AA9-AA9-AA99"
name="alphanumeric"
placeholder="____-___-____"
maxlength="13"

> type="text"

Reviewer: This code also differs from what is shown in the component code section. Instead of a full code example, could it just show an example of the related `mask` attribute?

```

{:.usa-content-list }
<h3>Additional parameters</h3>

{:.usa-content-list }
Additional parameters can be used to force uppercase or lowercase characters. Add `forceupper` or `forcelower` to the `input` element and set either of them to true to force this behavior.

{:.usa-content-list }

> <h3>Error handling</h3>

Reviewer: Having an error handling section without instructions seems to stand out from our other guidance pages. What if instead it was a note about customizing the error handling message (assuming that data attribute becomes a thing).
Related: https://designsystem.digital.gov/components/file-input/#using-the-file-input-component

{:.usa-content-list }
The input mask utilizes screen reader-friendly logic to gently notify a user if the character they typed is incorrect. The Preview section above illustrates this.

{:.usa-content-list }
<h3>Available format characters</h3>

{:.usa-content-list }
Here is a complete list of format characters that can be used in your implementation of input mask:

> {:.usa-content-list }

Reviewer: Can this table match the format of the other tables (for settings, variants, properties, etc) on our guidance pages? They use at least the `usa-table--borderless site-table-responsive site-table-simple` classes as well as `font-lang-3xs`, among others.

Author: @amyleadem check the latest for all the updates... Looks so much better, thanks for all suggestions!
https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/mm-input-mask-guidance/components/input-mask/

{:.usa-content-list }
The examples in the Preview section above are common input masks, but you can adapt the code to meet the needs of your project. In order to extend the functionality of the input mask, you must include a `placeholder` attribute as the visible input mask, using the underscore `_` to represent a character the user should enter. You may use punctuation literals like `[space]`, `-`, `/`, or `(` to help format your mask, but these literals will not be included in the final submitted value. _See a complete list of format characters below_.

{:.usa-content-list }
- **Add component classes.** The structure should include a base element with the class `usa-input-mask`. Inside of that base element there should be an input element with the class `usa-input-mask__field` and a message element (span or div) with the class `usa-input-mask__message`.

> - **Add a `mask` attribute to the input element.** The mask attribute is used to define the format of your mask. Use `9`s for numbers and `A`s for letters. For example, a typical SSN number format in the mask attribute should look like `999-99-9999`.

Reviewer: Does it make sense to list out all of the component attributes in a properties/procedures section like we do for components like [time picker](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/mm-input-mask-guidance/components/time-picker/)? 
Asking because there are a fair number of attributes to account for, and it is a little more clear to see them in a tabular format and grouped together.

Author: @amyleadem This is complete and is looking so much better! https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/mm-input-mask-guidance/components/input-mask/

- **Add a `placeholder` attribute to the input element.** This indicates the mask format. Use underscores `_` for character input placeholders combined with any valid format characters (see below) to customize your input mask. e.g. A US telephone `placeholder` attribute may look like `(___) ___-____`
- **Add a `maxlength` attribute to the input element.** This will be used as the limit referenced in the message and for validation.

> - **Set the attribute `type` to numeric if your mask is all numbers.** This will display the numeric keypad instead of the full keyboard.

Reviewer: Did you mean `inputmode` instead of `type`?
```suggestion
- **Set the attribute `inputmode` to numeric if your mask is all numbers.** This will display the numeric keypad instead of the full keyboard.
```

> - **Set the attribute `type` to numeric if your mask is all numbers.** This will display the numeric keypad instead of the full keyboard.

Reviewer: Also, should we give guidance for displaying a `tel` keypad?
